src/plugins/matrixops/softmax_operation.cpp

```diff
@@ -0,0 +1,188 @@
+// Copyright (c) 2018 Bita Hasheminezhad
+// Copyright (c) 2018 Hartmut Kaiser
+//
+// Distributed under the Boost Software License, Version 1.0. (See accompanying
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#include <phylanx/config.hpp>
+#include <phylanx/ir/node_data.hpp>
+#include <phylanx/plugins/matrixops/softmax_operation.hpp>
+#include <phylanx/util/matrix_iterators.hpp>
+
+#include <hpx/include/lcos.hpp>
+#include <hpx/include/naming.hpp>
+#include <hpx/include/util.hpp>
+#include <hpx/throw_exception.hpp>
+#include <hpx/util/iterator_facade.hpp>
+#include <hpx/util/optional.hpp>
+
+#include <algorithm>
+#include <cstddef>
+#include <cstdint>
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include <blaze/Math.h>
+
+///////////////////////////////////////////////////////////////////////////////
+namespace phylanx { namespace execution_tree { namespace primitives
+{
+    ///////////////////////////////////////////////////////////////////////////
+    match_pattern_type const softmax_operation::match_data =
+    {
+        hpx::util::make_tuple("softmax",
+        std::vector<std::string>{"softmax(_1)","softmax(_1,_2)"},
+        &create_softmax_operation, &create_primitive<softmax_operation>,
+        "a, axis\n"
+        "Args:\n"
+        "\n"
+        "    a (array_like) : input array\n"
+        "    axis (optional, integer): an axis to softmax along. The\n"
+        "        default is the last axis (axis == -1) of an array. Axis\n"
+        "        is effective for >1d arrays.\n"
+        "\n"
+        "Returns:\n"
+        "\n"
+        "Returns an array of the same shape which is the normalized exponential\n"
+        "function of the given array.  The resulting array consists of real\n"
+        "values in the range (0..1], which add up to 1 in direction of the given axis")
+    };
+
+    ///////////////////////////////////////////////////////////////////////////
+    softmax_operation::softmax_operation(primitive_arguments_type&& operands,
+        std::string const& name, std::string const& codename)
+      : primitive_component_base(std::move(operands), name, codename)
+    {}
+
+    primitive_argument_type softmax_operation::softmax0d() const
+    {
+        return primitive_argument_type{static_cast<double>(1.)};
+    }
+
+    primitive_argument_type softmax_operation::softmax1d(arg_type&& arg) const
+    {
+        auto v = arg.vector();
+        return primitive_argument_type{blaze::softmax(v)};
+    }
+
+    primitive_argument_type softmax_operation::softmax2d_axis0(
+        arg_type&& arg) const
+    {
+        auto m = arg.matrix();
+        blaze::DynamicMatrix<double> result(m.rows(), m.columns());
+        for (std::size_t i = 0; i < m.columns(); ++i)
+        {
+            column(result, i) = softmax(column(m, i));
```

Comment:
While `column` and `softmax` are being found via ADL, I'd suggest to explicitly decorate those as `blaze::column` and `blaze::softmax`.